contributors/design-proposals/preserve-order-in-strategic-merge-patch.md

```diff
@@ -0,0 +1,547 @@
+# Preserve Order in Strategic Merge Patch
+
+Author: @mengqiy
+
+## Motivation
+
+Background of the Strategic Merge Patch is convered [here](../devel/strategic-merge-patch.md).
+
+The Kubernetes API may apply semantic meaning to the ordering of items within a list,
+however the strategic merge patch does not keeping the ordering of elements.
+Ordering has semantic meaning for Environment variables,
+as later environment variables may reference earlier environment variables,
+but not the other way around.
+
+One use case is the environment variables. We don't preserve the order which causes
+issue [40373](https://github.com/kubernetes/kubernetes/issues/40373).
+
+## Proposed Change
+
+Changes are all in strategic merge patch package.
+The proposed solution is similar to the solution used for deleting elements from lists of primitives.
+
+Add to the current patch, a directive ($setElementOrder) containing a list of element keys -
```

Comment:
@adohe I think you mean that the ordering is not semantically important for every list, and should we add a comment tag to the field definition indicating whether order has semantic meaning.  (e.g. The containers list in a PodTemplate does not have semantically meaningful order).

I think this would be be  interesting to do in a follow up.  Since this is totally optional and sent by the client, the client should decide whether it wants to control the order or not for a given patch (yes it should want to do so)